jax_distribution/systems/rec_mappo_rware.py

```diff
@@ -0,0 +1,737 @@
+                    transition.reward,
+                )
+                delta = reward + config["gamma"] * next_value * (1 - done) - value
+                gae = delta + config["gamma"] * config["gae_lambda"] * (1 - done) * gae
+                return (gae, value), gae
+
+            _, advantages = jax.lax.scan(
+                _get_advantages,
+                (jnp.zeros_like(last_val), last_val),
+                traj_batch,
+                reverse=True,
+                unroll=16,
+            )
+            return advantages, advantages + traj_batch.value
+
+        advantages, targets = _calculate_gae(traj_batch, last_val)
+
+        def _update_epoch(update_state: Tuple, _: Any) -> Tuple:
+            """Update the network for a single epoch."""
+
+            def _update_minibatch(train_state: Tuple, batch_info: Tuple) -> Tuple:
+                """Update the network for a single minibatch."""
+                params, opt_state = train_state
+                (
+                    init_policy_hstate,
+                    init_critic_hstate,
+                    traj_batch,
+                    advantages,
+                    targets,
+                ) = batch_info
+
+                def _loss_fn(
+                    params: FrozenDict,
+                    opt_state: OptState,
+                    traj_batch: PPOTransition,
+                    gae: chex.Array,
+                    targets: chex.Array,
+                ) -> Tuple:
+                    """Calculate the loss."""
+                    # RERUN NETWORK
+                    _, actor_policy, value = apply_fn(
+                        params,
+                        (init_policy_hstate.squeeze(0), init_critic_hstate.squeeze(0)),
+                        (
+                            traj_batch.obs,
+                            traj_batch.done[:, :, 0],
+                        ),
+                    )
+                    log_prob = actor_policy.log_prob(traj_batch.action)
+
+                    # CALCULATE VALUE LOSS
+                    value_pred_clipped = traj_batch.value + (
+                        value - traj_batch.value
+                    ).clip(-config["clip_eps"], config["clip_eps"])
+                    value_losses = jnp.square(value - targets)
+                    value_losses_clipped = jnp.square(value_pred_clipped - targets)
+                    value_loss = (
+                        0.5 * jnp.maximum(value_losses, value_losses_clipped).mean()
+                    )
+
+                    # CALCULATE ACTOR LOSS
+                    ratio = jnp.exp(log_prob - traj_batch.log_prob)
+                    gae = (gae - gae.mean()) / (gae.std() + 1e-8)
+                    loss_actor1 = ratio * gae
+                    loss_actor2 = (
+                        jnp.clip(
+                            ratio,
+                            1.0 - config["clip_eps"],
+                            1.0 + config["clip_eps"],
+                        )
```

Comment:
Also only saw this now, would need to change all files, but might be better to pull this out into `clipped_ratio = ...` for readability

Reply:
We feel like this is already descriptive enough

Reply:
are you okay to reconsider